Common/include/geometry/dual_grid/CPoint.hpp

```diff
@@ -88,7 +89,14 @@ class CPoint {
   su2vector<bool> Agglomerate;                 /*!< \brief This flag indicates if the element has been agglomerated. */
 
   su2vector<unsigned short> nNeighbor;    /*!< \brief Number of neighbors, needed by some numerical methods. */
+
+  /*--- Closest element on a viscous wall, and distance to it. ---*/
   su2activevector Wall_Distance;          /*!< \brief Distance to the nearest wall. */
+  su2vector<int> ClosestWall_Rank; /*!< \brief Rank of process holding the closest wall element. */
+  su2vector<unsigned short> ClosestWall_Zone; /*!< \brief Zone index of closest wall element. */
```

Comment:
The code cites
 > Aupoix, B. and Spalart, P. R., "Extensions of the Spalart-Allmaras Turbulence Model to Account for Wall Roughness,"
 > International Journal of Heat and Fluid Flow, Vol. 24, 2003, pp. 454-462.
 > See https://turbmodels.larc.nasa.gov/spalart.html#sarough for detailed explanation.

The TMR page says that you require the " conventional Nikuradse sand roughness scale height" at the wall. In the old implementation each marker is allowed to have its specific roughness, and for integration at one point, the roughness of the closest marker is chosen. As I just wanted to fix the bug and I'm not familiar with the theory, I kept this level of generality.

I encountered the bug in setup with two fluid zones connected by a sliding mesh interface. 

The point is that in `CPhysicalGeometry::SetWallDistance`, the call to `WallADT::DetermineNearestElement` gives the closest wall marker in the `iZone` (which is not the zone in which `this CPhysicalGeometry` lives in, generally). It is therefore wrong to access the wall roughness array of `this CPhysicalGeometry` with the marker ID that belongs to a different zone. In my case, valgrind reported an invalid read, because there were different numbers of markers in both zones.

Maybe the dependency on the rank is irrelevant, I don't know if all markers have the same marker IDs in all ranks... I don't care ;)

Reply:
They don't. iMarker only makes sense locally and different ranks may have different boundaries.
If you try to do any MPI as part of a boundary condition you will have a bad time.